### PR TITLE
Netbox Inventory plugin enhancements

### PR DESCRIPTION
**Pull request by mkeetman**
##### SUMMARY
- Added the ability for requests for VM inventory information to collect config_context data issue #53393
- Added a function to extract dcim device interface information

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
Netbox inventory plugin

##### ADDITIONAL INFORMATION
The inventory extraction is quite expensive and slow.  The Netbox API doesn't appear to lend itself to easy access to this information (except requesting each interface individually)

Tested against Netbox v2.5.3

### CODE REVIEW

**Review by resmo — APPROVED** [review]
LGTM

File: lib/ansible/plugins/inventory/netbox.py [hunk 1]
```diff
@@ -39,6 +39,11 @@
                   region, site, role, platform, and/or tenant. Please check official netbox docs for more info.
             default: False
             type: boolean
+        interfaces:
+            description:
+                - If True, it adds dcim interface information in host vars.
+            default: False
+            type: boolean
```
Comment (resmo) [inline lib/ansible/plugins/inventory/netbox.py:46]: please add a `version_added: "2.8"`

Comment (mkeetman) [reply lib/ansible/plugins/inventory/netbox.py:46]: Thanks for the review.  Apologies for the subsequent commit mess

**Review by resmo — APPROVED** [review]
shipit

File: lib/ansible/plugins/inventory/netbox.py [hunk 2]
```diff
@@ -298,6 +308,15 @@ def extract_primary_ip6(self, host):
     def extract_tags(self, host):
         return host["tags"]
 
+    def extract_interfaces(self, host):
+        try:
+            if self.interfaces:
+                url = urljoin(self.api_endpoint, "/api/dcim/interfaces/?limit=0&device_id=%s" % (str(host["id"])))
```
Comment (bcoca) [inline lib/ansible/plugins/inventory/netbox.py]: use `to_text()` instead of `str()` as its py2/3 portable

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/inventory/netbox.py [hunk 3]
```diff
@@ -277,27 +287,36 @@ def extract_manufacturer(self, host):
...
         except Exception:
             return
 
     def extract_tags(self, host):
         return host["tags"]
 
+    def extract_interfaces(self, host):
+        try:
+            if self.interfaces:
+                url = urljoin(self.api_endpoint, "/api/dcim/interfaces/?limit=0&device_id=%s" % (to_text(host["id"])))
```
Comment (kb-light) [inline lib/ansible/plugins/inventory/netbox.py]: You do not differentiate between dcim and VMs, so this would raise a similar bug as #53393.

Comment (mkeetman) [reply lib/ansible/plugins/inventory/netbox.py]: This was a conscious decision, and probably short sighted.  I've included it.  It is admittedly not as useful in the context of a VM as it for network devices - I needed it originally to extract VLANs present on the network  device for configuration purposes.  IMO for this to be useful for VMs a secondary lookup should be executed to retrieve the associated IP details.  Can that be a future PR?

**Review by FragmentedPacket — APPROVED** [review]
LGTM

**Comment by FragmentedPacket** [comment]
shipit

**Comment by empi89** [comment]
Would you also like to add this for ip addresses? When dealing with networking configuration the ip addresses of the network interfaces which are stored in netbox are quite essential.

**Comment by mkeetman** [comment]
> Would you also like to add this for ip addresses? When dealing with networking configuration the ip addresses of the network interfaces which are stored in netbox are quite essential.

This makes perfect sense.  This information isn't returned by the DCIM API, so additional calls to the IPAM API would be required when "count_ipaddresses> 0".  

Thanks for the suggestion.

File: lib/ansible/plugins/inventory/netbox.py [hunk 4]
```diff
@@ -39,6 +39,12 @@
                   region, site, role, platform, and/or tenant. Please check official netbox docs for more info.
             default: False
             type: boolean
+        interfaces:
+            description:
+                - If True, it adds the device or virtual machine interface information in host vars.
+            default: False
+            type: boolean
+            version_added: "2.8"
```
Comment (resmo) [inline lib/ansible/plugins/inventory/netbox.py]: please adjust the version to `version_added: "2.9"`

Comment (mkeetman) [reply lib/ansible/plugins/inventory/netbox.py]: version_added bumped. Thanks for your review

**Comment by empi89** [comment]
@mkeetman Would suggest to do the ip address thing in a separate PR.

**Comment by FragmentedPacket** [comment]
Was this closed on accident?

**Comment by mkeetman** [comment]
No I am going to split it out into the two features and try again - sorry, had intended to do this basically immediately but have been distracted

**Pull request closed without merging**